src/clients/meta/MetaClient.cpp

```diff
@@ -3473,6 +3474,44 @@ folly::Future<StatusOr<bool>> MetaClient::ingest(GraphSpaceID spaceId) {
   };
   return folly::async(func);
 }
-
+bool MetaClient::loadSessions() {
+  auto session_list = listSessions().get();
+  if (!session_list.ok()) {
+    LOG(ERROR) << "List sessions failed, status:" << session_list.status();
+    return false;
+  }
+  decltype(sessionMap_) session_map;
+  decltype(killedPlans_) killed_plans;
+  for (auto& session : session_list.value().get_sessions()) {
+    session_map[session.get_session_id()] = session;
+    for (auto& query : session.get_queries()) {
+      if (query.second.get_status() == cpp2::QueryStatus::KILLING) {
+        killed_plans.insert({session.get_session_id(), query.first});
+      }
+    }
+  }
+  {
+    folly::RWSpinLock::WriteHolder holder(sessionLock_);
+    sessionMap_ = std::move(session_map);
+    killedPlans_ = std::move(killed_plans);
+  }
+  return true;
+}
+StatusOr<cpp2::Session> MetaClient::getSessionFromCache(const nebula::SessionID& session_id) {
+  if (!ready_) {
+    return Status::Error("Not ready!");
+  }
+  folly::RWSpinLock::ReadHolder holder(sessionLock_);
+  auto it = sessionMap_.find(session_id);
+  if (it != sessionMap_.end()) {
+    return it->second;
+  }
+  return Status::SessionNotFound();
+}
+bool MetaClient::checkIsPlanKilled(SessionID sessionId, ExecutionPlanID planId) {
+  // TODO(hs.zhang): Use RCU to avoid write blocking read
+  folly::RWSpinLock::ReadHolder holder(sessionLock_);
```

Comment:
The lock itself could be a bottleneck under concurrent scenario, especially you check it in every row. Maybe need to do some profiling test here. Consider only check for every N rows.

Besides, `std::set<std::pair<SessionID, ExecutionPlanID>>` may be a hot spot too, but it won't be a big issue if the check is not very frequent.

Reply:
Okay, I optimized the following points

1. Replace RWLock with RCU
2. Replace std::set with folly::F14FastSet
3. check for every N rows

Regarding the third point, there are two ways. The first is to judge whether it has reached N times outside the checkIsPlanKilled function. This method is accurate, but more complicated; the other is to define a static variable in the checkIsPlanKilled function to count uniformly. This method is simple to implement but It will not be very precise. Which do you prefer?

Reply:
Sorry for late response, the latter one is ok for now.